### PR TITLE
SOLR-14401: Track distrib/shard metrics differently

### PR DESCRIPTION
**Pull request by dsmiley**
* only do for SearchHandler, not all request handlers
* track all the same details at the shard level as request (more metrics)
* don't limit this to SolrCloud

https://issues.apache.org/jira/browse/SOLR-14401

_(I have some comments I'll post later)_
_(I have yet to updated affected tests; seems to be 1-2)_

### CODE REVIEW

**Review by HoustonPutman — COMMENTED** [review]
These changes look good to me. I don't have any particular opinion on the new naming, I see pros and cons, but if it works, it works.

Obviously we need some documentation around this, and I'm assuming that's coming next.

I think we need to change the default prometheus exporter config & graphana dashboard to use these new values, which will also give us a good indication if theres an issue with the way that the new naming is done.

File: solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java [hunk 1]
```diff
@@ -154,23 +138,50 @@ public SolrMetricsContext getSolrMetricsContext() {
...
+  /**
+   * Does this handler receive per-shard (or similar) requests contributing to larger request?
+   * This is used to track metrics separately.
+   * @see ShardParams#IS_SHARD
+   */
+  protected boolean supportsDistribedRequests() {
+    return false;
+  }
+
+  protected static class RhMetrics {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java:157]: What does Rh mean in this context?

File: solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java [hunk 2]
```diff
@@ -154,23 +138,50 @@ public SolrMetricsContext getSolrMetricsContext() {
...
+    static final RhMetrics NO_OP;
+    static {
+      var metricsConfig = new MetricsConfig.MetricsConfigBuilder().setEnabled(false).build();
+      var ctx = new SolrMetricsContext(new SolrMetricManager(null, metricsConfig), "dummy", "dummy");
+      NO_OP = new RhMetrics(ctx, "dummy", "dummy");
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java:162]: Why does path need two elements here?

**Review by madrob — COMMENTED** [review]
Do we need to consolidate this with AuthenticationPlugin metrics as well?

File: solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java (hunk 1 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java:157]: RequestHandler.
Alternatively I could call it simply `Metrics` but sometimes it's annoying when simple/common classes are used across the codebase as it shows up in import pickers if it's needed somewhere else.  It's no big deal; I'll just rename it to that.

**Comment by dsmiley** [comment]
Changes to the prometheus config are necessary.  I want to follow some how-to on use of Grafana we have to ensure it looks right.
Also, performance-statistics-reference.adoc

**Comment by dsmiley** [comment]
> Do we need to consolidate this with AuthenticationPlugin metrics as well?

Huh?

**Comment by dsmiley** [comment]
Definitely see JIRA for my overall comments.

I want to point out that I observed that the "totalTime" metric has been a nanosecond number in recent years, yet once upon a time it was milliseconds.  This change was very likely inadvertent.  Our prometheus solr-exporter-config.xml shows that it *thinks* it's milliseconds.  It's not; RequestHandlerBase increments this counter by "elapsed", the response of `timer.stop()` -- nanoseconds.  Years ago it had invoked `MetricUtils.nsToMs(` but it appears @sigram removed this as a part of other changes in 2017 sometime -- https://github.com/apache/solr/commit/d8df9f8c9963c2fc1718fd471316bf5d964125ba

**Comment by janhoy** [comment]
> Definitely see JIRA for my overall comments.
> 
> I want to point out that I observed that the "totalTime" metric has been a nanosecond number in recent years, yet once upon a time it was milliseconds. This change was very likely inadvertent. Our prometheus solr-exporter-config.xml shows that it _thinks_ it's milliseconds. It's not; RequestHandlerBase increments this counter by "elapsed", the response of `timer.stop()` -- nanoseconds. Years ago it had invoked `MetricUtils.nsToMs(` but it appears @sigram removed this as a part of other changes in 2017 sometime -- [d8df9f8](https://github.com/apache/solr/commit/d8df9f8c9963c2fc1718fd471316bf5d964125ba)

I have seen some very strange number in grafana dashboards for Solr with huge numbers. So I think this is a bug. Can you file a new JIRA for it?

**Comment by dsmiley** [comment]
https://issues.apache.org/jira/browse/SOLR-16073

**Comment by dsmiley** [comment]
Ready for review again.  Pending:
* updated CHANGES.txt, major-changes-in-solr-9.adoc
* precommit fix / merge with main
* updating the Solr ref guide

**Comment by dsmiley** [comment]
I think it's ready finally.  Please check out the ref guide explanation that I just pushed.

I think the outdated MetricsQueryTemplateTest ought to be replaced with a very different integration test that is more of a sanity check that certain metrics we expect to see are present.  Such a test could be done via docker and it'd thus offer more/better test coverage that other aspects are working too.

File: solr/prometheus-exporter/build.gradle [hunk 3]
```diff
@@ -61,6 +61,7 @@ task run(type: JavaExec) {
   main = project.ext.mainClass
   classpath = sourceSets.main.runtimeClasspath
   systemProperties = ["log4j.configurationFile":"file:conf/log4j2.xml"]
+  args = ["-f", "conf/solr-exporter-config.xml"]
```
Comment (HoustonPutman) [inline solr/prometheus-exporter/build.gradle:64]: Ahh we could also add this to the classpath above like we do in the bin script.

Though it's not something we need to do in this PR.

**Review by HoustonPutman — COMMENTED** [review]
This looks good to me, but I haven't gone through and tested it yet. Might get time to do that tomorrow

**Comment by HoustonPutman** [comment]
Ok found an issue, the `core-query` template isn't the only affected template for prometheus. `core` has metrics that contain `[shard]` as well, including (but not limited to):
- `solr_metrics_core_errors_total`
- `solr_metrics_core_requests_total `
- `solr_metrics_core_timeouts_total`

It was an easy fix, adding the same code from the `core-query` template, but it also means that  the `internal=false` tag will be added for other handlers that may be used internally, we just don't include the tag for it. Personally I think that's fine and it can be seen as a possible improvement for the future. (I don't think the graphana dashboard is necessarily impacted by this. We could certainly add the tag filter for these metrics, but I don't think its necessary for this PR)

I'll go ahead and push the fix and let you take a look @dsmiley. Otherwise I'm good with this.

**Comment by dsmiley** [comment]
Thanks Houston!  Yeah I wondered about the other templates but I didn't notice that they wold process such cases.

**Pull request merged**